src/RoboSaverModule.sol

```diff
@@ -1,16 +1,103 @@
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                       EVENTS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    event SafeTopup(address indexed safe, uint256 amount, uint256 timestamp);
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                     CONSTRUCTOR
+    //////////////////////////////////////////////////////////////////////////*/
+    constructor(address _delayModule, address _rolesModule, address _topupAgent) {
+        delayModule = IDelayModifier(_delayModule);
+        rolesModule = IRolesModifier(_rolesModule);
+
+        topupAgent = _topupAgent;
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                      MODIFIERS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @notice Checks whether a call is authorized to trigger top-up or exec queue txs
+    modifier onlyTopupAgents() {
+        if (msg.sender != topupAgent) revert NotTopupAgent(msg.sender);
+        _;
+    }
+
+    /// @dev Check condition and determine whether a task should be executed by Gelato.
+    function checker() external view returns (bool canExec, bytes memory execPayload) {
+        address cachedAvatar = delayModule.avatar();
+
+        uint256 queueNonce = delayModule.queueNonce();
+        uint256 txNonce = delayModule.txNonce();
+
+        if (txNonce != queueNonce) {
+            uint256 txQueuedAt = delayModule.getTxCreatedAt(queueNonce - 1);
+            // @note triggers the condition for exec the pendant tx in the delay module
+            if (block.timestamp - txQueuedAt >= delayModule.txCooldown()) {
+                return (true, abi.encodeWithSelector(IDelayModifier.executeNextTx.selector));
+            }
+        }
+
+        uint256 balance = EUR_E.balanceOf(cachedAvatar);
+        (, uint128 maxRefill,,,) = rolesModule.allowances(SET_ALLOWANCE_KEY);
+
+        // @note it will queue the tx for topup
+        if (balance < maxRefill) {
+            uint256 topupAmount = maxRefill - balance;
+            return (true, abi.encodeWithSelector(IRoboSaverModule.safeTopup.selector, cachedAvatar, topupAmount));
+        }
+
+        return (false, bytes("No gnosispay to topup"));
+    }
+
+    function safeTopup(address _avatar, uint256 _topupAmount) external onlyTopupAgents {
+        // @note logic for top-up is to be defined <> amm
+        emit SafeTopup(_avatar, _topupAmount, block.timestamp);
+    }
+
+    function execQueuedTransaction() external onlyTopupAgents {
+        // @note it will execute the pending tx in the delay module
+    }
 
-    constructor(address _delayModule) {
-        delayModule = IDelayModule(_delayModule);
+    function transferErc20(address _token, uint256 _tokenTransferAmount, address _destination) external {
```

Comment:
left it as external to have a quick way to visualise the full testing/trace flow till its replace w/ the real methods/actions the placeholder sections of `safeTopup` & `execQueuedTransaction`

ofc as is currently will not be in the final version of the sc, will be a security flaw otherwise